x-pack/plugins/security_solution/server/endpoint/services/actions/clients/endpoint/endpoint_actions_client.ts

```diff
@@ -0,0 +1,339 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import type { FleetActionRequest } from '@kbn/fleet-plugin/server/services/actions';
+import { v4 as uuidv4 } from 'uuid';
+import { ResponseActionsClientError } from '../errors';
+import { getActionRequestExpiration } from '../../create/write_action_to_indices';
+import { stringify } from '../../../../utils/stringify';
+import type { HapiReadableStream } from '../../../../../types';
+import type {
+  ResponseActionsApiCommandNames,
+  ResponseActionAgentType,
+} from '../../../../../../common/endpoint/service/response_actions/constants';
+import type {
+  ExecuteActionRequestBody,
+  GetProcessesRequestBody,
+  IsolationRouteRequestBody,
+  ResponseActionGetFileRequestBody,
+  UploadActionApiRequestBody,
+  ResponseActionsRequestBody,
+} from '../../../../../../common/api/endpoint';
+import { ResponseActionsClientImpl } from '../lib/base_response_actions_client';
+import type {
+  ActionDetails,
+  HostMetadata,
+  GetProcessesActionOutputContent,
+  KillOrSuspendProcessRequestBody,
+  KillProcessActionOutputContent,
+  ResponseActionExecuteOutputContent,
+  ResponseActionGetFileOutputContent,
+  ResponseActionGetFileParameters,
+  ResponseActionParametersWithPidOrEntityId,
+  ResponseActionsExecuteParameters,
+  ResponseActionUploadOutputContent,
+  ResponseActionUploadParameters,
+  SuspendProcessActionOutputContent,
+  LogsEndpointAction,
+  EndpointActionDataParameterTypes,
+} from '../../../../../../common/endpoint/types';
+import type { CommonResponseActionMethodOptions } from '../lib/types';
+import { DEFAULT_EXECUTE_ACTION_TIMEOUT } from '../../../../../../common/endpoint/service/response_actions/constants';
+
+export class EndpointActionsClient extends ResponseActionsClientImpl {
+  protected readonly agentType: ResponseActionAgentType = 'endpoint';
+
+  private async checkAgentIds(ids: string[]): Promise<{
+    valid: string[];
+    invalid: string[];
+    allValid: boolean;
+    hosts: HostMetadata[];
+  }> {
+    const foundEndpointHosts = await this.options.endpointService
+      .getEndpointMetadataService()
+      .getMetadataForEndpoints(this.options.esClient, [...new Set(ids)]);
```

Comment:
Q: Do we need to handle errors here? In case the metadata promise throws/returns an error?

Reply:
hmm. good point. Need to think about that more. Some errors we just can't recover from and for those, I don't think they should be "silent" and still allow (for automated actions) the creation of the (failed) action request